File: problems/3147_taking_maximum_energy_from_the_mystic_dungeon/3147_taking_maximum_energy_from_the_mystic_dungeon.java
class Solution {
    public int maximumEnergy(int[] energy, int k) {
        int n = energy.length;
        int[] dp = energy;
        int max = Integer.MIN_VALUE;
        for (int i = n - 1; i >= 0; i--) {
            if (i + k < n) {
                dp[i] += dp[i + k];
            }
            max = Math.max(dp[i], max);
        }
        return max;
    }
}

File: problems/3147_taking_maximum_energy_from_the_mystic_dungeon/3147_taking_maximum_energy_from_the_mystic_dungeon.py
class Solution:
    def maximumEnergy(self, energy: List[int], k: int) -> int:
        n = len(energy)
        dp = energy[:]
        for i in range(n-1, -1, -1):
            if i + k < n:
                dp[i] += dp[i + k]
        return max(dp)